### DISCUSSION TITLE
Guys, System.Buffers.Slice is so slow, it's supposed to interop with high performance System.IO.Pipelines

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/runtime#94629] [General] [answered]
it's supposed to interop with high performance System.IO.Pipelines which returns ReadOnlySegment that needs to be sliced
I spent two days using stackalloc in F# combating unverifiable codes to implement a module to vectorize data that are flowing through the pipeline, I ran profiler after doing optimizations and saw System.Buffers.Slice was taking 94% of the cpu

![image](https://github.com/dotnet/runtime/assets/35312807/5e9b1174-a9b7-422d-a542-47ef29a7c3b1)

I had to use slice because I need to get exactly the number of bytes needed for vector, but if 16MB of data takes 22 seconds to complete idk if System.IO.Pipelines is that high of a performance I was aiming at about 96GB/S

But turns out I achieved 1GB/S with 1GB of data and ~1000ms of execution by setting minimum segment size to 500MB
![image](https://github.com/dotnet/runtime/assets/35312807/56dbd603-4044-43fd-adc2-278188478eb6)

**Comment by tannergooding** [comment, marked as answer]
`ROSequence` tracks a sequence, which can possibly be many segments, and so slicing arbitrary buffers out of it can be quite expensive and is not trivial to inline. In the worst case, it ends up having to walk through many segments to find the relevant start (effectively walking a linked list) and then continue walking until it finds the end.

There are potentially things that could be done to improve that for the fast path, but you probably want to be using something like `SequenceReader`, which helps offset the costs (particularly when doing sequential reading)

CC. @GrabYourPitchforks

**Reply by Xyncgas** [reply]
Thanks for the insight your help is very useful
